### ISSUE TITLE
Figure out better maintenance model

### ISSUE THREAD

**Issue by ztellman** [clj-commons/aleph#441]
I've been struggling to give Aleph the attention it deserves.  I no longer use it in my day job, and am unlikely to do so in the near future.  @kachayev has been helping enormously, but some larger PRs have been languishing for months waiting for my review.  This is issue is to help myself, Oleksii, and any other interested parties figure out a better model.

A decent template for this, I think, is Netty.  Trustin Lee created the project, and put in most of the initial work, but over the past few years has allowed Norman Maurer to take the lead.  Given his contributions thus far, I'm happy to allow Oleksii to do the same.  I'd like to have input on the APIs for new features, to make sure there's some consistency across the project, but I'm worried that will become the new chokepoint for improving the project.  I'm interested to hear people's thoughts on this point.

For now I intend to directly maintain Manifold and Dirigiste, as they change more slowly, and the implementations are a bit less straightforward.  I am, however, happy to support anyone who's interested in working their way up to lead maintainer on these projects.

**Comment by gsnewmark** [comment]
I'm definitely biased as we work together with @kachayev, but still I support such decision in case @kachayev is ready to be more involved in the maintenance (of which I'm not exactly sure). 

A rough idea regarding the new features/APIs: should we consider something like the RFC process, when an interested party proposes new feature/API and then it's open for discussion for everyone with the final word/veto reserved for @ztellman? To ensure such RFCs don't stagnate but don't put unwanted excessive pressure on @ztellman at the same time, we can decide on some specific timeframe (e. g., half a year/year/etc.) after which RFC could be approved by other lead maintainer(s) in case @ztellman doesn't have time to check it personally.

**Comment by ztellman** [comment]
To be clear, I wasn't intending for @kachayev to take on more responsibility, but rather to simply not wait on my input/approval for everything.  If this seems like a burden, we can figure out another way to approach the problem.

**Comment by kachayev** [comment]
@ztellman Thanks once again for mentioning. And for the courage to bring this question up as a problem that needs to be addressed. Not a "burden" or something like this at all. I can take more responsibilities on the project support. But I have a few things that I want to outspeak/make clear before the conversation goes forward.

1. Recently my job responsibilities somewhat drifted away from communication systems to data processing, so now I spend less time with Aleph & Manifold than I did a few months ago. My hobby interests are somewhat related to both projects (mostly structured concurrency & async flows verifications). But that's still a hobby.

2. I've been working with Aleph, Manifold, and Dirigisty for a long time now within a pretty specific use case: large microservices system, hundreds of millions of HTTP requests per day, low latency and low overhead requirements, a pretty large team working with the same codebase. Meaning that I'm somewhat biased and have a strong opinion on what works and what doesn't which is obviously not universal and doesn't cover all users of libraries. E.g. I have a lot of doubts about using any implicit thread pools, I know that `let-flow` does more harm than good in the long run. Etc etc etc. You should also note "HTTP" in the description above. So, I don't work a lot with other protocols, like TCP or UDP or whatever.

Having said all of that...

All of my recent contribution to Aleph falls into one of two simple categories:

a) some features that my team lacked and those that we could not implement in a separate library (e.g. async DNS, websocket pings, idle cleanups, shutdown etc).

b) fixing incorrect behavior that either we faced in production or were reported or those I've caught reading the source code.

All of that was done without (or better to say, with a lack of) general vision, plan or any long term ideas where the project is going or where do we want to see it in future. That's the key question I want to clarify and I would love to hear any thoughts and feedback on this from you, Zach, and from anyone involved/interested in the development and popularization of the library. Apart from just fixing bugs & issues... from defining the global vision we can move on with a specific plan or at least a list of features that are missing or those we want to see in Aleph and other libraries.

Regarding the process... I don't think that complexity of RFCs is justified taking into account the number of people involved in the process right now. Later on, if that changes with time, we will justify, for sure. For now, I think we can focus on preparing the next stable release with all small fixes that were done recently and those that are still in PRs. The only one PR that I would suggest to postpone is #406 as it's non-trivial and might do more harm than I want to :)

I can also spend some time to improve test cases or even introduce some new approaches for testing, so we can be sure that the release is stable. Probably we should also extend benchmarking infrastructure to cover performance, memory consumption etc and keep track of any improvements or downgrades introduced with recent changes. But the most important... it would be really good to keep a list of libraries that are using Aleph and other libraries under the hood and a list of companies/developers who's willing to be early adopters and have abilities to test newest releases with their test suites or in production. To make that happen as earlier as possible, we can set a good tradition of making `-alpha*` release for any PR merged into the master branch. So... if you want to participate and you're reading this, - please reply in the comments :)

I have also a few small stabilization ideas for Aleph on my todo list (mostly around optimizations of Netty's usage). I will work on those, will try to fix earlier reported issues, will work on PR to Netty where it's necessary. But to make this more... meaningful in the long run, let's talk about future vision & bigger ideas. 

Thanks,

**Comment by ztellman** [comment]
To your initial points:

1. Aleph was something I got to use in my day job, but it was never the center of my responsibilities.  I've never, for instance, used `let-flow` in production.  A lot of the fancier aspects of Aleph and Manifold were theoretically useful, rather than a direct extension of my own work.  If there are people who are using these more seriously, I'd want them to be involved, but I don't see a problem with a hobbyist running the project.

2. This also mirrors my own applications of the library.  I know some people use Aleph because they think it's cool, but where Aleph is *necessary* are these high-throughput, low-latency cases.  I want to maintain good default behaviors, but the point of the library was always to give expert users all the escape hatches they need to completely control how things run.

There are a few things that represent ongoing work on this project: keeping up with Netty, and keeping up with `clj-http`.  There's a lot of Netty APIs which are not exposed (HTTP2, Stomp, etc), but I'm not sure those are worth adding in at this point, since most people don't even use the TCP and UDP functionality.

Broadly, I think there are three classes of new feature work:

1. Filling in gaps in the existing protocols (e.g. websocket pings, multipart encodings)
2. Operational improvements (e.g. graceful shutdown)
3. Performance improvements (e.g. changing how `chain` and `with-executor` interact)

I think 1 is the majority of the work, 2 will come up occasionally, and 3 likely involves Manifold rather than Aleph.  Assuming these categories seem reasonable to everyone, I think it would be useful to create tags for issues, and create issues for any future work people would like to see on the projects.

Again, I'm hoping people using Aleph in their day jobs will chime in here.  I honestly have no idea who's using Aleph at this point, and I want to make sure the decisions here reflect the needs of as many active users as possible.

**Comment by gsnewmark** [comment]
> I honestly have no idea who's using Aleph at this point, and I want to make sure the decisions here reflect the needs of as many active users as possible.

@ztellman @kachayev should we start with compiling a list of Aleph users to get at least a rough understanding of how and by whom Aleph is currently used (and possibly reach out directly to these users)? Something along the lines of https://github.com/tonsky/rum#whos-using-rum. I (and @kachayev) know a couple of companies using Aleph in our area, which could be a starting point.

**Comment by kachayev** [comment]
> I honestly have no idea who's using Aleph at this point, and I want to make sure the decisions here reflect the needs of as many active users as possible.

Same here ¯\_(ツ)_/¯ Would love to hear from other users!

> but where Aleph is necessary are these high-throughput, low-latency cases

I would also add here:

1. all use cases where we need something more than request/response, like streaming responses, WebSockets, HTTP2 etc. 

2. proxy servers, API gateways etc

Maybe that should be mentioned somewhere in docs/readme? 🤔

> keeping up with Netty, and keeping up with clj-http

Oh, yeah... I've been working on this for quite some time and still have a few things from Netty API I'm working on right now. Regarding `clj-http`, we're almost set with the pairing... all of the missing pieces are not that trivial, at least from what I could recall.

> Filling in gaps in the existing protocols

I can prepare a list of those gaps that could be filled in leveraging Netty functionality (like I did for websocket pings and multipart requests).

**Comment by alexander-yakushev** [comment]
We use Aleph in a couple of relatively busy HTTP services, the peak being 50k RPS / 4 Gbps per machine. As I started pushing Aleph to such loads I discovered several performance-related issues that I'd like to be fixed in the future. So, my primary interests in evolving Aleph/Manifold are performance, predictability, observability, and graceful degradation. A few concrete topics that come to mind:

- Minimizing the number of allocations/copying in HTTP request handler. byte-streams do that quite well already, but I think there is room for more improvement.
- Revising Manifold execution model to make it more robust (@kachayev had already suggested that).

As for the usecases, so far we use Aleph exclusively for req/resp HTTP communication, no websockets yet. We also moved to raw Netty for the few TCP usecases we had.

Given that, I'd be happy to contribute fixes and improvements as we're pushing Aleph usage further. I also think that @kachayev would make a great steward for the project.

**Comment by mccraigmccraig** [comment]
we ( yapster.info ) are using aleph for our api, via [juxt/yada](https://github.com/juxt/yada), and manifold just about everywhere on our backend

we don't use `let-flow` at all - we mostly use [cats/mlet](https://github.com/funcool/cats/blob/master/src/cats/core.cljc#L233) along with the [cats deferred context](https://github.com/funcool/cats/blob/master/src/cats/labs/manifold.clj) for composition, (and very occasionally [cats/alet](https://github.com/funcool/cats/blob/master/src/cats/core.cljc#L413) when we need graph-like composition)

in terms of the classifications of work - we don't really need anything more from aleph and manifold atm. keeping up with netty is important, performance improvements are always nice, and a more robust execution model is hard to argue with, but we have solutions (homegrown and from yada) for websocket pings and multipart encodings

**Comment by kachayev** [comment]
@mccraigmccraig  Thanks for mentioning `juxt/yada`, it's a good example of a library that leverages both Aleph and Manifold. I know they rolled out their own multipart requests processing some time ago as it was not covered by Aleph. I'm not sure tho' how active the most recent development of `yada` is ... at least the version of Aleph they're using is quite old.

@alexander-yakushev 50k RPS sound impressing. Allocations within HTTP request handler is something I've been looking into for a while, we can discuss this topic separately. It feels like a lot could be done there to cover more high-throughput use cases. Execution model robustness is an extensive topic by itself. Would be glad to put some effort into it even tho' Manifold is a way more sophisticated library comparing to Aleph. I'm curious about one thing... when you decided to use Aleph to handle 50k RPS, what other options were on the table?

P.S. Regarding `let-flow`, @ztellman @mccraigmccraig I've just mentioned this one as a showy example of being opinionated on some things :)

**Comment by alexander-yakushev** [comment]
> I'm curious about one thing... when you decided to use Aleph to handle 50k RPS, what other options were on the table?
In that particular case we previously used http-kit and just ran more machines for the service. But having fewer machines is more convenient operationally.

**Comment by kachayev** [comment]
@ztellman What do you think about exposing HTTP2? There was [an issue](https://github.com/ztellman/aleph/issues/251) on the topic that got some attention (read, number of thumbs up is quite high) but it was more than 2 years ago. I'm not sure who much of the real interest it has in the Clojure community, but it might be a case of a double-edged sword problem: a lack of instruments dictates a lack attempt to use it. Personally, I think we would probably at least try to leverage some theoretical benefits of HTTP2 if we had a good way to use this protocol in our systems (I mean w/o being forced to use plain Netty handlers).

**Comment by kachayev** [comment]
@ztellman I've just created 2 new labels, reviewed a few older issues and opened a few newer (something that I'm working on right now). We also have a few pretty old issues that I would mark as "support request", but I'm not sure if it's still relevant after many months:

* https://github.com/ztellman/aleph/issues/301
* https://github.com/ztellman/aleph/issues/316
* https://github.com/ztellman/aleph/issues/372
* https://github.com/ztellman/aleph/issues/324

Also not sure if this issue was addressed in Dirigiste: https://github.com/ztellman/aleph/issues/235

**Comment by ztellman** [comment]
I'm going to be traveling for the rest of December, and will have spotty internet access.  I've opened #450 to get a sense of who's using Aleph, which will hopefully inform some of the questions I raised here.

**Comment by kachayev** [comment]
@ztellman I [tried](https://twitter.com/kachayev/status/1071111936664838144) to start a conversation on extending Ring's spec, with basically no luck. We can actually do both:

1. HTTP2 as underlying implementation detail on how to run the same clients and servers. It still makes sense to use w/o any proxy, e.g. for inter-server communication or for a client that talks to an HTTP2-powered API.  

2. Explicit API to handle stream within established connection as a manifold steam for advanced use cases (the same way it works for WebSockets). Specific HTTP2 features might be exposed here, like server-side push or connection priorities. I don't think that there's a problem of not being covered by the ring spec, WebSocket protocol is not covered either. We can come up with some specification to provide some level of unification and portability between libraries, but right now there are not that many choices of what and where to port.

**Comment by SevereOverfl0w** [comment]
>  I'm not sure tho' how active the most recent development of yada is ... at least the version of Aleph they're using is quite old.

We use yada on all our client projects.  Development has slowed a little over the last year as we've been focusing on other parts of our stack.  We're currently doing some new development on yada, and expect it to see a lot of attention over the next year.

**Comment by jimrthy** [comment]
I'm a late-comer to the discussion, and currently just a hobbyist (so don't fit the criteria for  https://github.com/ztellman/aleph/issues/450 but wanted to chime in somewhere).

First: I really appreciate all the time and effort that's gone into creating a low-latency, high-throughput library on top of Netty, because I really don't want to cope with all the ugly details.

I've spent the past couple of years of my spare time translating CurveCP into clojure on top of aleph. That's all about the udp (https://github.com/jimrthy/frereth-cp). Aleph and manifold solved some serious problem I had trying to get this to work (I think those stemmed from trying to mix core.async with agents).

I've recently been working on a log handling GUI to debug problems with that (https://github.com/jimrthy/frereth/tree/master/apps/log-viewer). That brings in web sockets (I'm very happy that I can't tell you the first thing about any of the details). And that means that it just made sense to use aleph as my underlying web server.

I happened to discover this thread last night because I was starting to do some research into implementing an http/2 microservice at my day job. I'll pitch the idea to the engineering team on Monday morning. HTTP/2 *is* the company future for internal APIs, and my team currently has a very short time window before we just ditch higher level languages (in the form of python)  in favor of raw java (because that's what we know as an enterprise company).
